distributed/tests/test_core.py

```diff
@@ -839,6 +842,34 @@ async def connect_to_server():
         assert all(t.cancelled() for t in tasks)
 
 
+@gen_test()
+async def test_remove_cancels_connect_attempts():
+    loop = asyncio.get_running_loop()
+    connect_started = asyncio.Event()
+    connect_finished = loop.create_future()
+
+    async def connect(*args, **kwargs):
+        connect_started.set()
+        await connect_finished
+
+    async def connect_to_server():
+        with pytest.raises(CommClosedError, match="Address removed."):
+            await rpc.connect("tcp://0.0.0.0")
+
+    async def remove_address():
+        await connect_started.wait()
+        rpc.remove("tcp://0.0.0.0")
+
+    rpc = await ConnectionPool(limit=1)
+
+    with mock.patch("distributed.core.connect", connect):
```

Comment:
Instead of mocking, you could also just connect to a server with either a very slow or blocked listener or one that is not replying on the handshake

Reply:
I thought mocking was cleaner here, I'll look into using a slow server

Reply:
also the other tests monkeypatch with `monkeypatch.setitem(backends, "tcp", SlowBackend())`, so I still think using `mock.patch` on the `connect` function is cleanest here

Reply:
when testing this:

```python
@gen_test()
async def test_remove_cancels_connect_attempts():
    loop = asyncio.get_running_loop()
    connect_started = asyncio.Event()
    connect_finished = asyncio.Event()

    class BrokenHandshakeListener(TCPListener):
        async def on_connection(self, comm):
            try:
                connect_started.set()
                await comm.read(1)
            finally:
                connect_finished.set()

    async with BrokenHandshakeListener(
        address="tcp://",
        comm_handler=lambda: None,
    ) as listener:
        rpc = await ConnectionPool(limit=1)

        async def connect_to_server():
            with pytest.raises(CommClosedError, match="Address removed."):
                await rpc.connect(listener.contact_address)

        async def remove_address():
            await connect_started.wait()
            rpc.remove(listener.contact_address)

        await asyncio.gather(
            connect_to_server(),
            remove_address(),
        )

        await connect_finished.wait()
```

I hit a race condition in this asyncio.wait_for where the cancellation is ignored:
https://github.com/dask/distributed/blob/9ddee125e23f38aa337ec3c6b3cc99184196d916/distributed/comm/core.py#L291-L295
https://github.com/python/cpython/blob/924a3bfa28578802eb9ca77a66fb5d4762a62f14/Lib/asyncio/tasks.py#L472

this is because the localhost connect completes immediately and the cancellation arrives just as the `connect() -> asyncio.wait_for()` coroutine is about to be resumed, this is very unlikely in production because the code will be waiting in `comm.read()` or `asyncio.sleep(backoff)` and can be resolved in the test by adding an `asyncio.sleep(0.5)` before calling `rpc.remove()`:


```python
        async def remove_address():
            await connect_started.wait()
            await asyncio.sleep(0.5)  # avoid issuing a .cancel() after connect but before comm.read()
            rpc.remove(listener.contact_address)
```

or do you think it's best to leave this with a mocked connect function?

Reply:
I think we should come back to this once https://github.com/dask/distributed/pull/7571 is done. Triggering this edge case is interesting and I believe we've encountered this a couple of times in the past (in CI).

No need for any action on this PR.